internal/provider/waypoint/resource_waypoint_action.go

```diff
@@ -0,0 +1,629 @@
+	var methodEnum waypoint_models.HashicorpCloudWaypointActionConfigFlavorCustomMethod
+	switch method {
+	case "GET":
+		methodEnum = waypoint_models.HashicorpCloudWaypointActionConfigFlavorCustomMethodGET
+	case "POST":
+		methodEnum = waypoint_models.HashicorpCloudWaypointActionConfigFlavorCustomMethodPOST
+	case "PUT":
+		methodEnum = waypoint_models.HashicorpCloudWaypointActionConfigFlavorCustomMethodPUT
+	case "PATCH":
+		methodEnum = waypoint_models.HashicorpCloudWaypointActionConfigFlavorCustomMethodPATCH
+	case "DELETE":
+		methodEnum = waypoint_models.HashicorpCloudWaypointActionConfigFlavorCustomMethodDELETE
+	default:
+		return methodEnum, fmt.Errorf("unknown method")
+	}
+	return methodEnum, nil
+}
+
+func (r *ActionResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_waypoint_action"
+}
+
+func (r *ActionResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		// This description is used by the documentation generator and the language server.
+		MarkdownDescription: "The Waypoint Action resource managed the lifecycle of an Action.",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Description: "The ID of the Action.",
+				Computed:    true,
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the Action.",
+				Required:    true,
+			},
+			"project_id": schema.StringAttribute{
+				Description: "The ID of the HCP project where the Action is located.",
+				Computed:    true,
+				Optional:    true,
+			},
+			"organization_id": schema.StringAttribute{
+				Description: "The ID of the HCP organization where the Action is located.",
+				Computed:    true,
+			},
+			// An Action description must be fewer than 125 characters if set.
+			"description": schema.StringAttribute{
+				Description: "A description of the Action.",
+				Optional:    true,
+			},
+			"namespace_id": schema.StringAttribute{
+				Description: "Internal Namespace ID",
+				Computed:    true,
+			},
+			// Action URL is required if the action is custom
+			"action_url": schema.StringAttribute{
+				Description: "The URL to trigger an action on. Only used in custom mode",
+				Optional:    true,
+			},
+			"request": schema.SingleNestedAttribute{
+				Description: "The kind of HTTP request this should trigger.",
+				Required:    true,
+				Attributes: map[string]schema.Attribute{
+					"custom": schema.SingleNestedAttribute{
+						Description: "Custom mode allows users to define the HTTP method, the request body, etc.",
+						Optional:    true,
+						Attributes: map[string]schema.Attribute{
+							"method": schema.StringAttribute{
+								Description: "The HTTP method to use for the request.",
+								Required:    true,
```

Comment:
`schema.StringAttribute` might have a way to say like "always upper-case this value" or something?

Reply:
I found it! There is a built-in `Validators` field for every attribute!